service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/PertinentInformationObservationValueMapper.java

```diff
@@ -200,4 +217,144 @@ private static boolean isDenominatorPresent(Ratio ratio) {
+            .stream()
+            .map(this::buildComponentPertinentInformation)
+            .filter(StringUtils::isNotEmpty)
+            .collect(Collectors.joining(StringUtils.SPACE));
+
+        return String.format(COMPONENT_TEMPLATE, componentPertinentInformation);
+    }
+
+    private String buildComponentPertinentInformation(Observation.ObservationComponentComponent component) {
+        List<String> componentPertinentInformationList = List.of(
+            extractComponentCode(component),
+            extractComponentValue(component),
+            extractComponentInterpretationCode(component),
+            extractComponentInterpretationText(component),
+            extractComponentReferenceRange(component)
+        );
+
+        var componentText = componentPertinentInformationList.stream()
+            .filter(StringUtils::isNotEmpty)
+            .collect(Collectors.joining(StringUtils.SPACE));
+
+        if (StringUtils.isNotEmpty(componentText)) {
+            return String.format(COMPONENT_DELIMITER_TEMPLATE, componentText);
+        }
+
+        return componentText;
+    }
+
+    private String extractComponentCode(Observation.ObservationComponentComponent component) {
+        if (component.hasCode()) {
+            var code = CodeableConceptMappingUtils.extractTextOrCoding(component.getCode());
+            if (code.isPresent()) {
+                return String.format(COMPONENT_CODE_TEMPLATE, code.get());
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private String extractComponentValue(Observation.ObservationComponentComponent component) {
+        if (component.hasValue()) {
+            var value = component.getValue();
+
+            if (COMPONENT_VALUE_FUNCTIONS.containsKey(value.getClass())) {
+                return COMPONENT_VALUE_FUNCTIONS.get(value.getClass()).apply(value);
+            } else {
+                return mapObservationValueToPertinentInformation(value);
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private String extractComponentInterpretationCode(Observation.ObservationComponentComponent component) {
+        if (component.hasInterpretation()) {
+            if (component.getInterpretation().hasCoding()) {
+                var userSelectedCoding = component
+                    .getInterpretation()
+                    .getCoding()
+                    .stream()
+                    .filter(Coding::hasUserSelected)
+                    .findFirst();
+
+                if (userSelectedCoding.isPresent()) {
+                    return String.format(COMPONENT_INTERPRETATION_CODE_TEMPLATE, userSelectedCoding.get().getCode());
+                } else {
+                    return String.format(COMPONENT_INTERPRETATION_CODE_TEMPLATE,
+                        component.getInterpretation().getCodingFirstRep().getCode());
+                }
+            }
+        }
+        return StringUtils.EMPTY;
```

Comment:
The two nested `if`s can be combined, and there's some optimisation we can make to the coding selection, I think:

```suggestion
        if (component.hasInterpretation() && component.getInterpretation().hasCoding()) {
            var userSelectedCoding = component
                .getInterpretation()
                .getCoding()
                .stream()
                .filter(Coding::hasUserSelected)
                .findFirst();
            
            var codingTouse = userSelectedCoding.orElseGet(() -> component.getInterpretation().getCodingFirstRep());
            
            return String.format(COMPONENT_INTERPRETATION_CODE_TEMPLATE, codingTouse.getCode());
        }
        return StringUtils.EMPTY;
```